client/db/src/lib.rs

```diff
@@ -2245,13 +2250,17 @@ impl<Block: BlockT> sc_client_api::backend::Backend<Block> for Backend<Block> {
 
 		match self.blockchain.header_metadata(hash) {
 			Ok(ref hdr) => {
-				if !self.have_state_at(&hash, hdr.number) {
-					return Err(sp_blockchain::Error::UnknownBlock(format!(
-						"State already discarded for {:?}",
-						block
-					)))
-				}
-				if let Ok(()) = self.storage.state_db.pin(&hash) {
+				// NOTE: similar to `sp_state_machine::Storage::get` but we can't use
+				// `sp_state_machine::Storage::get` directly in order to avoid dead lock
```

Comment:
This is intended to avoid the locking here:
https://github.com/paritytech/substrate/blob/1fd71c7845d6c28c532795ec79106d959dd1fe30/client/state-db/src/lib.rs#L590

because when `hint` is called within `StateDbSync::pin` we were already holding this lock. 

But indeed, we may not need to query the noncanonical layer because it may already be checked before `hint` is called:
https://github.com/paritytech/substrate/blob/1fd71c7845d6c28c532795ec79106d959dd1fe30/client/state-db/src/lib.rs#L410